src/generic/stage1/utf8_validator.h

```diff
@@ -0,0 +1,29 @@
+namespace stage1 {
+/**
+ * Validates that the string is actual UTF-8.
+ */
+template<class checker>
+bool utf8_validate(const uint8_t * input, size_t length) {
+    checker c{};
+    buf_block_reader<64> reader(input, length);
+    while (reader.has_full_block()) {
+      simd::simd8x64<uint8_t> in(reader.full_block());
+      c.check_next_input(in);
+      reader.advance();
+    }
+
+    if (likely(reader.has_remainder())) {
```

Comment:
Note: this check is really just checking whether the file is empty. Might be better to just do that at the beginning. The way we broke things out, any non-empty file will *always* have a 1-64-byte block at the end (because of lenminusstep).

(Things should work out OK, I think, even if the file *is* empty, we'll just do a little extra work--so it's also possible to just remove the if statement entirely.)

Reply:
Ok.